From dd615814cc54815ada4883e4099cd86996bdcf2c Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:39:28 -0500
Subject: [drm] i915/dp: only use training pattern 3 on platforms that support it

Message-id: <1418852380-13061-78-git-send-email-rclark@redhat.com>
Patchwork-id: 102197
O-Subject: [RHEL7 drm 77/89] drm/i915/dp: only use training pattern 3 on platforms that support it
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since drm-intel-fixes-2014-10-30

commit 7809a61176b385ebb3299ea43c58b1bb31ffb8c0

Author:     Jani Nikula <jani.nikula@intel.com>
AuthorDate: Wed Oct 29 11:03:26 2014 +0200
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Mon Dec 15 14:28:42 2014 -0500

    drm/i915/dp: only use training pattern 3 on platforms that support it

    Ivybridge + 30" monitor prints a drm error on every modeset, since IVB
    doesn't support DP3 we should even bother trying to use it.

    This regression has been introduced in

    commit 06ea66b6bb445043dc25a9626254d5c130093199
    Author: Todd Previte <tprevite@gmail.com>
    Date:   Mon Jan 20 10:19:39 2014 -0700

        drm/i915: Enable 5.4Ghz (HBR2) link rate for Displayport 1.2-capable
    devices

    Reported-by: Dave Airlie <airlied@redhat.com>
    Reference: http://mid.gmane.org/1414566170-9868-1-git-send-email-airlied@gmail.com
    Cc: Todd Previte <tprevite@gmail.com>
    Cc: stable@vger.kernel.org (3.15+)
    Reviewed-by: Ville Syrjala <ville.syrjala@linux.intel.com>
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>

Signed-off-by: Rob Clark <rclark@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index a679a4e..61c370e 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -3210,9 +3210,10 @@ intel_dp_get_dpcd(struct intel_dp *intel_dp)
   }
  }
 
- /* Training Pattern 3 support */
+ /* Training Pattern 3 support, both source and sink */
  if (intel_dp->dpcd[DP_DPCD_REV] >= 0x12 &&
-     intel_dp->dpcd[DP_MAX_LANE_COUNT] & DP_TPS3_SUPPORTED) {
+     intel_dp->dpcd[DP_MAX_LANE_COUNT] & DP_TPS3_SUPPORTED &&
+     (IS_HASWELL(dev) || INTEL_INFO(dev)->gen >= 8)) {
   intel_dp->use_tps3 = true;
   DRM_DEBUG_KMS("Displayport TPS3 supported");
  } else
-- 
1.7.1